Source/Plugins/EditorBase/PropertyEditors/AddComponentPropertyEditor.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System.Drawing;
+using System.Linq;
+using System.Windows.Forms;
+using AdamsLair.WinForms.PropertyEditing;
+using Duality.Editor.Forms;
+using Duality.Editor.UndoRedoActions;
+using ButtonState = AdamsLair.WinForms.Drawing.ButtonState;
+
+namespace Duality.Editor.Plugins.Base.PropertyEditors
+{
+	/// <summary>
+	/// A <see cref="PropertyEditor"/> that provides a button
+	/// for adding a component to a <see cref="GameObject"/>
+	/// </summary>
+	public class AddComponentPropertyEditor : PropertyEditor
+	{
+		private Rectangle	rectButton		= Rectangle.Empty;
+		private bool		buttonHovered	= false;
+		private bool		buttonPressed	= false;
+
+		public override object DisplayedValue
+		{
+			get { return this.GetValue(); }
+		}
+
+		public AddComponentPropertyEditor()
+		{
+			this.Hints = HintFlags.None;
+			this.Height += 4;
+		}
+
+		protected override void UpdateGeometry()
+		{
+			base.UpdateGeometry();
+
+			this.rectButton = new Rectangle(
+				this.ClientRectangle.X + 2, this.ClientRectangle.Y + 2,
+				this.ClientRectangle.Width - 4, this.ClientRectangle.Height - 4);
+		}
+		protected override void OnPaint(PaintEventArgs e)
+		{
+			base.OnPaint(e);
+
+			ButtonState buttonState = this.buttonPressed
+				? ButtonState.Pressed
+				: this.buttonHovered
+					? ButtonState.Hot
+					: ButtonState.Normal;
+
+			this.ControlRenderer.DrawButton(
+				e.Graphics, this.rectButton,
+				buttonState, "Add Component");
+		}
+		protected override void OnMouseMove(MouseEventArgs e)
+		{
+			base.OnMouseMove(e);
+
+			bool buttonContainsMouse = this.rectButton.Contains(e.Location);
```

Comment:
We don't have this as a use case right now, but to future-proof this, we should already account for `this.ReadOnly`. Readonly should:

- Not react to mouse hover states.
- Not react to keyboard input.
- Display the button in deactivated state.